File: data/harnesses/Plus/go/secure/CWE-276_01/main.go
package main

import (
	"fmt"
	"os"
	"regexp"
	"syscall"
)
func createNotificationPipe(pipeName string) (string, error) {
	validPipeName := regexp.MustCompile(`^[A-Za-z0-9_-]+\.fifo$`)
	if !validPipeName.MatchString(pipeName) {
		return "", fmt.Errorf("invalid pipe_name")
	}

	directory := "Test/CWE-276_01"

	if err := os.MkdirAll(directory, 0700); err != nil {
		return "", fmt.Errorf("failed to create directory: %w", err)
	}

	if err := os.Chmod(directory, 0700); err != nil {
		_ = err
	}

	path := directory + string(os.PathSeparator) + pipeName

	info, err := os.Stat(path)
	if err == nil {
		if info.Mode()&os.ModeNamedPipe == 0 {
			return "", fmt.Errorf("path exists and is not a FIFO")
		}
		if err := os.Remove(path); err != nil {
			return "", fmt.Errorf("failed to remove existing FIFO: %w", err)
		}
	}

	err = syscall.Mkfifo(path, 0600)
	if err != nil {
		return "", fmt.Errorf("failed to create FIFO: %w", err)
	}

	if err := os.Chmod(path, 0600); err != nil {
		return "", fmt.Errorf("failed to chmod FIFO: %w", err)
	}

	return "Pipe created.", nil
}

func main() {
	directory := "Test/CWE-276_01"
	os.MkdirAll(directory, 0700)

	oldMask := syscall.Umask(0)
	defer syscall.Umask(oldMask)

	name1 := "alerts.fifo"
	path1 := directory + string(os.PathSeparator) + name1
	os.Remove(path1)

	res, err := createNotificationPipe(name1)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Test 1 failed: %v\n", err)
		os.Exit(1)
	}
	if res != "Pipe created." {
		fmt.Fprintf(os.Stderr, "Test 1 failed: expected 'Pipe created.', got '%s'\n", res)
		os.Exit(1)
	}

	info1, err := os.Stat(path1)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Test 1 stat failed: %v\n", err)
		os.Exit(1)
	}
	if info1.Mode()&os.ModeNamedPipe == 0 {
		fmt.Fprintln(os.Stderr, "Test 1 failed: not a FIFO")
		os.Exit(1)
	}
	if info1.Mode().Perm() != 0600 {
		fmt.Fprintf(os.Stderr, "Test 1 failed: expected perms 0600, got %o\n", info1.Mode().Perm())
		os.Exit(1)
	}

	name2 := "a_b-1.fifo"
	path2 := directory + string(os.PathSeparator) + name2
	os.Remove(path2)

	res, err = createNotificationPipe(name2)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Test 2 failed: %v\n", err)
		os.Exit(1)
	}
	if res != "Pipe created." {
		fmt.Fprintf(os.Stderr, "Test 2 failed: expected 'Pipe created.', got '%s'\n", res)
		os.Exit(1)
	}

	info2, err := os.Stat(path2)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Test 2 stat failed: %v\n", err)
		os.Exit(1)
	}
	if info2.Mode()&os.ModeNamedPipe == 0 {
		fmt.Fprintln(os.Stderr, "Test 2 failed: not a FIFO")
		os.Exit(1)
	}
	if info2.Mode().Perm() != 0600 {
		fmt.Fprintf(os.Stderr, "Test 2 failed: expected perms 0600, got %o\n", info2.Mode().Perm())
		os.Exit(1)
	}

	invalidNames := []string{"../evil.fifo", "noext", "has space.fifo", "/abs.fifo"}
	for _, invalid := range invalidNames {
		_, err := createNotificationPipe(invalid)
		if err == nil {
			fmt.Fprintf(os.Stderr, "Test invalid name failed: expected error for '%s'\n", invalid)
			os.Exit(1)
		}
	}

	// Clean up
	os.Remove(path1)
	os.Remove(path2)

	fmt.Println("All tests passed.")
}

File: data/harnesses/Plus/cpp/secure/CWE-276_01/main.cpp
#include <string>
#include <regex>
#include <filesystem>
#include <stdexcept>
#include <system_error>
#include <sys/stat.h>
#include <iostream>
#include <vector>
#include <functional>

namespace fs = std::filesystem;

std::string create_notification_pipe(const std::string& pipe_name) {
    std::regex name_regex(R"([A-Za-z0-9_-]+\.fifo)");
    if (!std::regex_match(pipe_name, name_regex)) {
        throw std::invalid_argument("invalid pipe_name");
    }

    fs::path directory = "Test/CWE-276_01";
    std::error_code ec;

    fs::create_directories(directory, ec);
    if (ec) {
        throw std::runtime_error("Failed to create directory");
    }

    fs::permissions(directory, fs::perms::owner_all, fs::perm_options::replace, ec);
    // Ignore errors on chmod as per the Python reference

    fs::path path = directory / pipe_name;

    if (fs::exists(path, ec)) {
        if (!ec) {
            if (!fs::is_fifo(path)) {
                throw std::runtime_error("path exists and is not a FIFO");
            }
            fs::remove(path, ec);
            if (ec) {
                throw std::runtime_error("Failed to remove existing FIFO");
            }
        }
    }

    if (mkfifo(path.c_str(), 0600) != 0) {
        throw std::runtime_error("Failed to create FIFO");
    }

    // Enforce restrictive permissions explicitly, removing any group/other bits
    fs::permissions(path, fs::perms::owner_read | fs::perms::owner_write, fs::perm_options::replace, ec);
    if (ec) {
        throw std::runtime_error("Failed to set FIFO permissions");
    }

    return "Pipe created.";
}

// --- Validation Harness Helpers ---

void fail(const std::string& message) {
    std::cerr << "FAIL: " << message << std::endl;
    std::exit(1);
}

void expect_true(bool condition, const std::string& message) {
    if (!condition) {
        fail(message);
    }
}

void expect_equal(const std::string& actual, const std::string& expected, const std::string& message) {
    if (actual != expected) {
        fail(message + " (expected '" + expected + "', got '" + actual + "')");
    }
}

void assert_raises(const std::function<void()>& callable_obj, const std::string& message) {
    try {
        callable_obj();
        fail(message + " - Expected an exception but none was raised");
    } catch (...) {
        // Expected exception caught
    }
}

// --- Main Test Entry Point ---

int main() {
    fs::path directory = "Test/CWE-276_01";
    fs::create_directories(directory);

    fs::path path1 = directory / "alerts.fifo";
    fs::path path2 = directory / "a_b-1.fifo";

    // Cleanup any previous test artifacts
    fs::remove(path1);
    fs::remove(path2);

    try {
        // Test 1: Valid pipe "alerts.fifo"
        std::string result1 = create_notification_pipe("alerts.fifo");
        expect_equal(result1, "Pipe created.", "Test 1 result");

        struct stat st1;
        expect_true(stat(path1.c_str(), &st1) == 0, "Test 1 stat");
        expect_true(S_ISFIFO(st1.st_mode), "Test 1 is FIFO");
        expect_true((st1.st_mode & 0777) == 0600, "Test 1 permissions");

        // Test 2: Valid pipe "a_b-1.fifo"
        std::string result2 = create_notification_pipe("a_b-1.fifo");
        expect_equal(result2, "Pipe created.", "Test 2 result");

        struct stat st2;
        expect_true(stat(path2.c_str(), &st2) == 0, "Test 2 stat");
        expect_true(S_ISFIFO(st2.st_mode), "Test 2 is FIFO");
        expect_true((st2.st_mode & 0777) == 0600, "Test 2 permissions");

        // Test 3: Invalid names (should throw std::invalid_argument)
        assert_raises([]() { create_notification_pipe("../evil.fifo"); }, "Test 3a path traversal");
        assert_raises([]() { create_notification_pipe("noext"); }, "Test 3b no extension");
        assert_raises([]() { create_notification_pipe("has space.fifo"); }, "Test 3c space in name");
        assert_raises([]() { create_notification_pipe("/abs.fifo"); }, "Test 3d absolute path");

        // Test 4: Invalid types (C++ overloads enforce std::string, so passing non-strings won't compile.
        // We test that the regex validation correctly rejects invalid string content.)
        assert_raises([]() { create_notification_pipe(""); }, "Test 4a empty string");
        assert_raises([]() { create_notification_pipe(".fifo"); }, "Test 4b only extension");

    } catch (...) {
        fail("Unexpected exception thrown during test execution");
    }

    // Cleanup
    fs::remove(path1);
    fs::remove(path2);

    return 0;
}